### PR TITLE
Manually trigger large Index Merges

### PR DESCRIPTION
**Pull request by riccardone**
Implemented new admin/mergeindexes http endpoint to trigger large Index Merge manually
Added a new AutomergeIndexes boolean option to deactivate the automatic index merges (default: true)

Fixes https://github.com/EventStore/EventStore/issues/1738

To test this PR 
1) set the new option AutomergeIndexes: False or --automerge-indexes=false
2) flood the test database with millions of test events
3) verify that the indexes are not automatically merged
4) make an http POST to the new endpoint: http://youreventstore:2113/admin/mergeindexes
5) verify that the indexes are merged

### CODE REVIEW

**Comment by shaan1337** [comment]
@riccardone hello, can you please rebase your changes against `master`? There will be a few conflicts to resolve (particularly in IndexMap.cs)

File: src/EventStore.ClusterNode/ClusterNodeOptions.cs [hunk 1]
```diff
@@ -242,6 +242,9 @@ public class ClusterNodeOptions : IOptions
 
         [ArgDescription(Opts.StructuredLogDescr, Opts.DbGroup)]
         public bool StructuredLog { get; set; }
+		
+		[ArgDescription(Opts.AutomergeIndexesDescr, Opts.DbGroup)]
+        public bool AutomergeIndexes { get; set; }
```
Comment (shaan1337) [inline src/EventStore.ClusterNode/ClusterNodeOptions.cs]: I would suggest `ManualIndexMerge` with a default of `False` instead of `AutoIndexMerge` so that the intent is clearer when setting it to `True`

File: src/EventStore.Core/Cluster/Settings/ClusterVNodeSettings.cs [hunk 2]
```diff
@@ -79,8 +79,10 @@ public class ClusterVNodeSettings
         public readonly bool SkipIndexScanOnReads;
         public readonly bool ReduceFileCachePressure;
         public readonly int InitializationThreads;
+        public readonly bool AutoMergeIndexes;
+
         public readonly bool GossipOnSingleNode;
-        public readonly bool FaultOutOfOrderProjections;
+		public readonly bool FaultOutOfOrderProjections;
```
Comment (shaan1337) [inline src/EventStore.Core/Cluster/Settings/ClusterVNodeSettings.cs]: fix indentation

File: src/EventStore.Core/Cluster/Settings/ClusterVNodeSettings.cs [hunk 3]
```diff
@@ -147,7 +149,8 @@ public class ClusterVNodeSettings
                                     bool reduceFileCachePressure = false,
                                     int initializationThreads = 1,
                                     bool faultOutOfOrderProjections = false,
-                                    bool structuredLog = false)
+                                    bool structuredLog = false,
+									bool autoMergeIndexes = true)
```
Comment (shaan1337) [inline src/EventStore.Core/Cluster/Settings/ClusterVNodeSettings.cs]: fix indentation

File: src/EventStore.Core/Index/IndexMap.cs [hunk 4]
```diff
@@ -412,19 +413,25 @@ public void SaveToFile(string filename)
...
+
+                        AddTableToTables(tables, level + 1, mergedTable);
+                        toDelete.AddRange(tables[level]);
+                        tables[level].Clear();
+                    }
                 }
             }
-
+            else
+                Log.Info("AutoMergeIndexes option is set to false and index-merge's must be manually triggered.");
```
Comment (shaan1337) [inline src/EventStore.Core/Index/IndexMap.cs]: option name must also be changed here if going ahead with `ManualIndexMerge`

File: src/EventStore.Core/ClusterVNode.cs [hunk 5]
```diff
@@ -508,7 +510,8 @@ protected virtual void OnNodeStatusChanged(VNodeStatusChangeArgs e)
                                                         !vNodeSettings.DisableScavengeMerging,
                                                         unsafeIgnoreHardDeletes: vNodeSettings.UnsafeIgnoreHardDeletes);
 
-			// ReSharper disable RedundantTypeArgumentsOfMethod
+            // ReSharper disable RedundantTypeArgumentsOfMethod
+            _mainBus.Subscribe<ClientMessage.MergeIndexes>(storageScavenger);
```
Comment (shaan1337) [inline src/EventStore.Core/ClusterVNode.cs]: It would be better if `ClientMessage.MergeIndexes` message is handled in `IndexCommitterService.cs` which is responsible for index-related stuff. The storage scavenger is responsible mainly for scavenging.

File: src/EventStore.Core/Services/Storage/StorageScavenger.cs [hunk 6]
```diff
@@ -52,6 +56,18 @@ public void Handle(SystemMessage.StateChangeMessage message)
             }
         }
 
+        public void Handle(ClientMessage.MergeIndexes message)
```
Comment (shaan1337) [inline src/EventStore.Core/Services/Storage/StorageScavenger.cs]: As mentioned [here](https://github.com/EventStore/EventStore/pull/1786/files#r244915589), it would be better to move this to `IndexCommiterService.cs` instead of `StorageScavenger.cs`

File: src/EventStore.Core/TransactionLog/Chunks/TFChunkScavenger.cs [hunk 7]
```diff
@@ -76,6 +76,21 @@ private IEnumerable<TFChunk.TFChunk> GetAllChunks(int startFromChunk)
             }
         }
 
+        public Task MergeIndexes()
```
Comment (shaan1337) [inline src/EventStore.Core/TransactionLog/Chunks/TFChunkScavenger.cs]: As mentioned [here](https://github.com/EventStore/EventStore/pull/1786/files#r244915589), it would be better to move this to `IndexCommiterService.cs` instead of `TFChunkScavenger.cs`

File: src/EventStore.Core/VNodeBuilder.cs [hunk 8]
```diff
@@ -1454,7 +1466,8 @@ public ClusterVNode Build(IOptions options = null, IPersistentSubscriptionConsum
                     _skipIndexScanOnReads,
                     _reduceFileCachePressure,
                     _initializationThreads,
-                    _faultOutOfOrderProjections);
+                    _faultOutOfOrderProjections,
+					_autoMergeIndexes);
```
Comment (shaan1337) [inline src/EventStore.Core/VNodeBuilder.cs]: fix indentation

File: src/EventStore.Core/Index/TableIndex.cs [hunk 9]
```diff
@@ -222,39 +226,49 @@ public void AddEntries(long commitPos, IList<IndexKey> entries)
...
+        }
 
-                lock (_awaitingTablesLock)
-                {
-                    var newTables = new List<TableItem> { new TableItem(_memTableFactory(), -1, -1) };
-                    newTables.AddRange(_awaitingMemTables.Select(
-                        (x, i) => i == 0 ? new TableItem(x.Table, prepareCheckpoint, commitPos) : x));
+        public void MergeIndexes()
+        {
+            TryProcessAwaitingTables(_indexMap.CommitCheckpoint, _indexMap.PrepareCheckpoint, true);
```
Comment (shaan1337) [inline src/EventStore.Core/Index/TableIndex.cs]: It is not correct to call `TryProcessAwaitingTables(long commitPos, long prepareCheckpoint, bool mergeIndexes)` here. The function should be called only when a MemTable is completed.

**Fix:**
`TryProcessAwaitingTables(bool mergeIndexes)` should directly be called.

**Issues:**
1. It will create a new Memtable everytime even if the previous one has not been completed. (e.g. it will even dump a MemTable with 0 index entries to a PTable)
2. The supplied values: `_indexMap.CommitCheckpoint` and `_indexMap.PrepareCheckpoint` are not correct and causes a crash when calling `/admin/mergeindexes` on an empty DB:

```
[21019,08,06:41:13.795] Error in TableIndex.ReadOffQueue
EXCEPTION OCCURRED
prepareCheckpoint should be non negative.
Parameter name: prepareCheckpoint
[21019,08,06:41:13.799] Global Unhandled Exception occurred.
EXCEPTION OCCURRED
prepareCheckpoint should be non negative.
Parameter name: prepareCheckpoint
[ERROR] FATAL UNHANDLED EXCEPTION: System.ArgumentOutOfRangeException: prepareCheckpoint should be non negative.
Parameter name: prepareCheckpoint
  at EventStore.Common.Utils.Ensure.Nonnegative (System.Int64 number, System.String argumentName) [0x00011] in <82749b1b247d4718bba6fcc8de345a00>:0 
  at EventStore.Core.Index.IndexMap.AddPTable (EventStore.Core.Index.PTable tableToAdd, System.Int64 prepareCheckpoint, System.Int64 commitCheckpoint, System.Func`3[T1,T2,TResult] upgradeHash, System.Func`2[T,TResult] existsAt, System.Func`2[T,TResult] recordExistsAt, EventStore.Core.Index.IIndexFilenameProvider filenameProvider, System.Byte version, System.Int32 indexCacheDepth, System.Boolean skipIndexVerify, System.Boolean mergeIndexes) [0x00000] in <3aba2e5e6d364a458c766d7dd95c2f8f>:0 
  at EventStore.Core.Index.TableIndex.ReadOffQueue (System.Boolean mergeIndexes) [0x00288] in <3aba2e5e6d364a458c766d7dd95c2f8f>:0 
  at EventStore.Core.Index.TableIndex+<>c__DisplayClass41_0.<TryProcessAwaitingTables>b__0 (System.Object x) [0x00000] in <3aba2e5e6d364a458c766d7dd95c2f8f>:0 
  at System.Threading.QueueUserWorkItemCallback.WaitCallback_Context (System.Object state) <0x7ff388e56140 + 0x0002a> in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Threading.ExecutionContext.RunInternal (System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state, System.Boolean preserveSyncCtx) <0x7ff388e50ee0 + 0x001a2> in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Threading.QueueUserWorkItemCallback.System.Threading.IThreadPoolWorkItem.ExecuteWorkItem () <0x7ff388e560f0 + 0x00025> in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Threading.ThreadPoolWorkQueue.Dispatch () [0x00074] in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
  at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback () <0x7ff388e55f70 + 0x00005> in <0f8aeac9d63d4b8aa575761bb4e65b79>:0 
```

Comment (shaan1337) [reply src/EventStore.Core/Index/TableIndex.cs]: Also:
1. If a background index merge/scavenge task is already running, the merge will not take place and user will not be informed. A check can be done on `_backgroundRunning = True` to log something and return. (this is particularly possible if you call `admin/mergeindexes` multiple times within a short amount of time)

2. Improvement: if `ManualIndexMerge` is set to `False`, the functionality can be disabled and the function can just log something and return (to be on the safe side for users who have not enabled this functionality)

File: src/EventStore.ClusterNode/ClusterNodeOptions.cs (hunk 1 above)
Comment (riccardone) [reply src/EventStore.ClusterNode/ClusterNodeOptions.cs]: AutoIndexMerge=True seems to be clearer and it represents the current behavior. 
I got that option name from Greg from the initial discussion (check the related issue) and I prefer it. Can we double check another opinion on that before I make all the changes?

File: src/EventStore.Core/Index/TableIndex.cs (hunk 9 above)
Comment (riccardone) [reply src/EventStore.Core/Index/TableIndex.cs]: Manually trigger the merge operation requires to pass through the 
TryProcessAwaitingTables(long commitPos, long prepareCheckpoint, bool mergeIndexes) method. I added a filter to avoid the exception.
1. Now it's fixed. If you send multiple http commands you'll get a log saying that the operation is running.
2. Even when the AutoMergeIndexes is set to True (normal behaviour) it is possible to manually trigger the Index Merge operation. This is a feature and not having it is not an improvement.

**Comment by riccardone** [comment]
@shaan1337  as discussed the option is called: AutoMergeIndexes (default true) and it will not be renamed
check the latest changes and comments to your previous messages to complete the review

File: src/EventStore.Core/VNodeBuilder.cs (hunk 8 above)
Comment (riccardone) [reply src/EventStore.Core/VNodeBuilder.cs]: I can't see the indentation issues in my editor

File: src/EventStore.Core/Services/Storage/IndexCommitterService.cs [hunk 10]
```diff
@@ -440,6 +447,41 @@ public bool IsReplicated(int commitCount)
...
+            if (!IsAllowed(message.User, message.CorrelationId, message.Envelope)) return;
+            if (_mergeIndexesTask != null && _mergeIndexesTask.Status == TaskStatus.Running)
+            {
+                Log.Info("Index Merge Operation already running...");
+                MakeReplyForMergeIndexes(message);
+                return;
+            }
+            _mergeIndexesTask = _tableIndex.MergeIndexes().ContinueWith(t =>
+            {
+                _tableIndex.Scavenge(_logManager.CreateLog(), default(CancellationToken));
```
Comment (shaan1337) [inline src/EventStore.Core/Services/Storage/IndexCommitterService.cs]: @riccardone Index merges automatically remove deleted entries in the PTables being merged. The intent of the `admin/mergeindexes` endpoint is only to merge pending PTables. I don't think it's a good idea to scavenge the index here since it's an operation that needs to go over the whole database (all chunk files and PTabls) and will be very time consuming/disk intensive. Besides, DB/Index scavenges can be launched manually by the user and if we want a user to be able to launch only an index scavenge (and not DB scavenge), a better place to add it would be in the `/admin/scavenge` endpoint in another PR.

File: src/EventStore.Core/ClusterVNode.cs [hunk 11]
```diff
@@ -226,12 +228,18 @@ protected virtual void OnNodeStatusChanged(VNodeStatusChangeArgs e)
             _mainBus.Subscribe<SystemMessage.BecomeShutdown>(storageReader);
             monitoringRequestBus.Subscribe<MonitoringMessage.InternalStatsRequest>(storageReader);
 
-            var indexCommitterService = new IndexCommitterService(readIndex.IndexCommitter, _mainQueue, db.Config.ReplicationCheckpoint, db.Config.WriterCheckpoint, vNodeSettings.CommitAckCount);
+            // IO DISPATCHER
+            var ioDispatcher = new IODispatcher(_mainQueue, new PublishEnvelope(_mainQueue));
+            var scavengerLogManager = new TFChunkScavengerLogManager(_nodeInfo.ExternalHttp.ToString(), TimeSpan.FromDays(vNodeSettings.ScavengeHistoryMaxAge), ioDispatcher);
+
+            var indexCommitterService = new IndexCommitterService(readIndex.IndexCommitter, _mainQueue,
+                db.Config.ReplicationCheckpoint, db.Config.WriterCheckpoint, vNodeSettings.CommitAckCount, tableIndex, scavengerLogManager);
```
Comment (shaan1337) [inline src/EventStore.Core/ClusterVNode.cs]: These changes need to be reverted. See: https://github.com/EventStore/EventStore/pull/1786#pullrequestreview-189685299

File: src/EventStore.Core/Services/Storage/IndexCommitterService.cs [hunk 12]
```diff
@@ -440,6 +447,41 @@ public bool IsReplicated(int commitCount)
...
+        {
+            message.Envelope.ReplyWith(new ClientMessage.MergeIndexesResponse(message.CorrelationId,
+                ClientMessage.MergeIndexesResponse.MergeIndexesResult.Started));
+        }
+
+        private bool IsAllowed(IPrincipal user, Guid correlationId, IEnvelope envelope)
+        {
+            if (user == null || (!user.IsInRole(SystemRoles.Admins) && !user.IsInRole(SystemRoles.Operations)))
+            {
+                envelope.ReplyWith(new ClientMessage.ScavengeDatabaseResponse(correlationId, ClientMessage.ScavengeDatabaseResponse.ScavengeResult.Unauthorized, null));
```
Comment (shaan1337) [inline src/EventStore.Core/Services/Storage/IndexCommitterService.cs]: @riccardone The reply message: `ClientMessage.ScavengeDatabaseResponse` seems incorrect. This `IsAllowed` check was also not present here previously.   Please don't change unrelated code between reviews since it's difficult to test and follow the changes.

File: src/EventStore.Core/Services/Storage/IndexCommitterService.cs (hunk 10 above)
Comment (riccardone) [reply src/EventStore.Core/Services/Storage/IndexCommitterService.cs]: It's already done and merged from the Scavenge_Indexes PR. When you scavenge database chunks using /admin/scavenge you also scavenge the indexes. That scavenge operation was running automatically before that Scavenge_Indexes PR as part of the index merge. Now that we are making the operation on the indexes controllable, I think that it make sense run Merge and Scavenge together on indexes.

Comment (shaan1337) [reply src/EventStore.Core/Services/Storage/IndexCommitterService.cs]: `That scavenge operation was running automatically before that Scavenge_Indexes PR as part of the index merge`

@riccardone Only the PTables being merged were being scavenged before the Scavenge_Indexes PR. e.g. if you had two 1M PTables being merged into a single 2M PTable and there were X deleted entries, the merge would result in a 2M-X PTable. All other PTables would remain unaffected and would not be scavenged.

File: src/EventStore.Core/Services/Storage/IndexCommitterService.cs (hunk 12 above)
Comment (riccardone) [reply src/EventStore.Core/Services/Storage/IndexCommitterService.cs]: ?? you asked to move logic to IndexCommitterService. Please don't ask to move logic is you don't want the correlated dependecies (libs, logc, methods, check) to be moved

Comment (riccardone) [reply src/EventStore.Core/Services/Storage/IndexCommitterService.cs]: The IsAllowed method that was checking permissions is now removed. The permission check for the IndexMerge operation is now only performed in the http controller

File: src/EventStore.Core/Index/TableIndex.cs [hunk 13]
```diff
@@ -249,6 +249,11 @@ public Task MergeIndexes()
             }, CancellationToken.None); 
         }
 
+        public bool IsBackgroungTaskRunning
```
Comment (bartelink) [inline src/EventStore.Core/Index/TableIndex.cs]: typo

Comment (riccardone) [reply src/EventStore.Core/Index/TableIndex.cs]: @bartelink fixed, thanks
